sound/core/compress_offload.c

```diff
@@ -488,6 +488,48 @@ snd_compr_get_codec_caps(struct snd_compr_stream *stream, unsigned long arg)
 }
 #endif /* !COMPR_CODEC_CAPS_OVERFLOW */
 
+int snd_compr_malloc_pages(struct snd_compr_stream *stream, size_t size)
+{
+	struct snd_dma_buffer *dmab;
+	int ret;
+
+	if (PCM_RUNTIME_CHECK(stream))
+		return -EINVAL;
+	dmab = kzalloc(sizeof(*dmab), GFP_KERNEL);
+	if (!dmab)
+		return -ENOMEM;
+	dmab->dev = stream->dma_buffer.dev;
+	ret = snd_dma_alloc_pages(dmab->dev.type, dmab->dev.dev, size, dmab);
+	if (ret < 0) {
+		kfree(dmab);
+		return ret;
+	}
+
+	snd_compr_set_runtime_buffer(stream, dmab);
+	stream->runtime->dma_bytes = size;
+	return 1;
+}
+EXPORT_SYMBOL(snd_compr_malloc_pages);
+
+int snd_compr_free_pages(struct snd_compr_stream *stream)
+{
+	struct snd_compr_runtime *runtime = stream->runtime;
+
+	if (PCM_RUNTIME_CHECK(stream))
+		return -EINVAL;
+	if (runtime->dma_area == NULL)
+		return 0;
+	if (runtime->dma_buffer_p != &stream->dma_buffer) {
+		/* It's a newly allocated buffer. Release it now. */
+		snd_dma_free_pages(runtime->dma_buffer_p);
+		kfree(runtime->dma_buffer_p);
+	}
+
+	snd_compr_set_runtime_buffer(stream, NULL);
+	return 0;
+}
+EXPORT_SYMBOL(snd_compr_free_pages);
```

Comment:
sorry, I don't get the point of these routines. We've used this compressed driver for 7+ years now, why do we need this memory allocation.
I must be missing the concept.

Reply:
DMA buffer is needed for BDL setup.

Qcom already broke the chain with: qdsp6 (_dai_compr_open). However, rather than adding common code, allocation has been inlined. I'd like to avoid that path. In consequence, existing usage (qdsp6) could be unified later on.

In bright future, I'd see pcm and compr coexisting tightly - very tightly. Many operations can be shared by both, and double implementation could be prevented (right now, main difference is: s/pcm/compr/g). That task is not within the scope for probes though.

Reply:
I need to spend more time on this, I just don't get how we added this for HDaudio  and we didn't need it before.